src/main/scala/com/microsoft/hyperspace/index/CoveringIndex.scala

```diff
@@ -0,0 +1,300 @@
+/*
+ * Copyright (2021) The Hyperspace Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.index
+
+import com.fasterxml.jackson.annotation.{JsonGetter, JsonRawValue}
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize
+import org.apache.spark.sql.{DataFrame, SaveMode}
+import org.apache.spark.sql.catalyst.catalog.BucketSpec
+import org.apache.spark.sql.functions.{col, input_file_name}
+import org.apache.spark.sql.hyperspace.utils.StructTypeUtils
+import org.apache.spark.sql.types.StructType
+
+import com.microsoft.hyperspace.HyperspaceException
+import com.microsoft.hyperspace.index.DataFrameWriterExtensions.Bucketizer
+import com.microsoft.hyperspace.shim.StructTypeConverter
+import com.microsoft.hyperspace.util.ResolverUtils
+import com.microsoft.hyperspace.util.ResolverUtils.ResolvedColumn
+
+case class CoveringIndex(
+    override val indexedColumns: Seq[String],
+    includedColumns: Seq[String],
+    // See schemaJson for more information about the annotation.
+    @JsonDeserialize(converter = classOf[StructTypeConverter.T]) schema: StructType,
```

Comment:
Can we just use schemaString, as discussed in #456? Other than this, LGTM! + please rebase onto master.

Reply:
+1 if @clee704 is also fine with it :)

Reply:
Note that those macros are optional for this to work. Do you still prefer schemaString over just schema?

It seems schemaString has been used instead of just schema because of the Jackson serialization issue. Shouldn't we fix the issue in the right way, instead of working around it? In the entire code base, the string is constantly being converted to `StructType` with `DataType.fromJson`. With a few simple annotations, this is not necessary. Having a string in the field and converting it to a typed object every time it is accessed seems unusual and raises my eyebrows whenever I think about it. There should be other reasons than the Jackson issue to justify it.

I admit the practical gain might not be great, but it costs almost nothing. Also, however small, the gain is on the user's side, whereas the cost is on our side. Since the user base is orders of magnitude larger than us developing Hyperspace, shouldn't we prioritize them?

We already have many shim files, and they don't contribute much to the overall complexity of the project as they are independent of each other and we can inspect them individually. It's like linear complexity is cheap compared to quadratic or exponential complexities, which correspond to other parts of the codebase.

Reply:
How about `Relation.dataSchemaJson`? Do we want to make it consistent? I would let @sezruby make the final decision.

Reply:
+1 for `Relation.dataSchemaJson`? I'm okay w/ Shim + StructType if you prefer that.

Reply:
Let's handle that in another PR as it is not related.